components/IR/Transceiver.cpp

```diff
@@ -0,0 +1,211 @@
+    receiveCfg.signal_range_max_ns = 12000000; // the longest duration for NEC signal is 9000us, 12000000ns > 9000us, the receive won't stop early
+    rmt_symbol_word_t buffer[mPacketSize];
+    ESP_ERROR_CHECK(rmt_receive(mRxCh, &buffer, sizeof(buffer), &receiveCfg));
+
+    rmt_rx_done_event_data_t eventToProccess;
+    while (true)
+    {
+        static constexpr auto msToWaitforVal = 1000;
+
+        if (xQueueReceive(mRxQueue, &eventToProccess, msToWaitforVal / portTICK_PERIOD_MS))
+        {
+            auto result = mNecParser.Parse(eventToProccess);
+            if (result.has_value())
+            {
+                auto [addr, data, isRepeat] = result.value();
+                mDataReceivedHandler(addr, data, isRepeat);
+            }
+            ESP_ERROR_CHECK(rmt_receive(mRxCh, &buffer, sizeof(buffer), &receiveCfg));
+        }
+        else
+        {
+            // ESP_LOGI(LOG_TAG, "No Symbols Received");
+            ESP_ERROR_CHECK(rmt_receive(mRxCh, &buffer, sizeof(buffer), &receiveCfg));
+        }
+    }
+}
+
+void Transceiver::enableRx()
+{
+    if (mIsRxEnabled)
+    {
+        return;
+    } // early return if we are already ready to Receive
+    rmt_enable(mRxCh);
+    xTaskCreate(this->receiveTaskImpl, RX_QUEUE_TASK_NAME, 4096, this, 5, &mReceiveProccess);
+    mIsRxEnabled = true;
+};
+
+void Transceiver::disableRx()
+{
+    if (!mIsRxEnabled)
+    {
+        return;
+    } // early return if we are already are disabled
+    rmt_disable(mRxCh);
+    vTaskDelete(mReceiveProccess);
+    mIsRxEnabled = false;
+};
+
+void Transceiver::enableTx()
+{
+    if (mIsTxEnabled)
+    {
+        return;
+    }
+    ESP_ERROR_CHECK(rmt_enable(mTxCh));
+    ir_nec_encoder_config_t nec_encoder_cfg = {
+        .resolution = Transceiver::IR_RESOLUTION_HZ,
+    };
+    ESP_ERROR_CHECK(rmt_new_ir_nec_encoder(&nec_encoder_cfg, &mNecEncoder));
+    mIsTxEnabled = true;
+};
+
+void Transceiver::disableTx()
+{
+    if (!mIsTxEnabled)
+    {
+        return;
+    }
+    rmt_disable(mTxCh);
+    if (mNecEncoder)
+    {
+        rmt_del_encoder(mNecEncoder);
+    }
+    mIsTxEnabled = false;
+};
+
+void Transceiver::send(uint16_t address, uint16_t data)
+{
+    if (!mIsTxEnabled)
+    {
+        ESP_LOGE(LOG_TAG, "Must enableTx() before using send()");
+        return;
+    }
+    rmt_transmit_config_t aTransmitConfig;
+    aTransmitConfig.loop_count = 0; // Todo this has to be zero
+    aTransmitConfig.flags.eot_level = 1;
+
+    const ir_nec_scan_code_t scan_code = {
+        .address = address,
+        .command = data,
+    };
+    ESP_ERROR_CHECK(rmt_transmit(mTxCh, mNecEncoder, &scan_code, sizeof(scan_code), &aTransmitConfig));
+    ESP_LOGI(LOG_TAG, "Sent - address:%d data:%c)", address, data);
+    // This teardown and setup feels like a bug but its the only way I could get it to send properly.
+    teardownTxChannel();
```

Comment:
I Really hate that this tears down and sets up the channel every time but if I don't do this it just does not send anything... any ideas are appreciated.